core/src/main/java/com/linecorp/armeria/internal/server/FileAggregatedMultipart.java

```diff
@@ -102,19 +107,43 @@ public static CompletableFuture<FileAggregatedMultipart> aggregateMultipart(Serv
     }
 
     private static CompletableFuture<Path> moveFile(Path file, Path targetDirectory,
-                                                    ExecutorService blockingExecutorService) {
+                                                    ExecutorService blockingExecutorService,
+                                                    ServiceRequestContext ctx) {
         return CompletableFuture.supplyAsync(() -> {
             try {
                 Files.createDirectories(targetDirectory);
                 // Avoid name duplication, create new file at target place and replace it.
-                return Files.move(file, Files.createTempFile(targetDirectory, null, ".multipart"),
-                                  StandardCopyOption.REPLACE_EXISTING);
+                final Path tempFile = createRemovableTempFile(targetDirectory, blockingExecutorService, ctx);
+                return Files.move(file, tempFile, StandardCopyOption.REPLACE_EXISTING);
             } catch (IOException e) {
                 throw new UncheckedIOException(e);
             }
         }, blockingExecutorService);
     }
 
+    private static Path createRemovableTempFile(Path targetDirectory,
+                                                ExecutorService blockingExecutorService,
+                                                ServiceRequestContext ctx) throws IOException {
+        final Path tempFile = Files.createTempFile(targetDirectory, null, ".multipart");
+        switch (ctx.config().multipartRemovalStrategy()) {
+            case NEVER:
+                break;
+            case ON_RESPONSE_COMPLETION:
+                ctx.log().whenComplete().thenAcceptAsync(unused -> {
+                    try {
+                        Files.deleteIfExists(tempFile);
+                    } catch (IOException e) {
+                        logger.warn("Failed to delete a temporary file: {}", tempFile, e);
+                    }
+                }, blockingExecutorService);
+                break;
+            case ON_JVM_SHUTDOWN:
+                tempFile.toFile().deleteOnExit();
```

Comment:
Wouldn't it result in a kind of resource leak, since the JVM needs to keep the list of files to remove on exit?

What do you think about this:

1. Introduce an additional flag `purgeMultipartUploadsOnExit` (default: true) that registers a shutdown hook that scans the directory and remove the items.
2. Remove `ON_JVM_SHUTDOWN` from `(default)multipartRemovalStrategy`.

Reply:
> Introduce an additional flag purgeMultipartUploadsOnExit (default: true) 

I think `purgeMultipartUploadsOnExit` and `MultipartRemovalStrategy` are mutually exclusive. If `MultipartRemovalStrategy.ON_RESPONSE_COMPLETION` is set, `purgeMultipartUploadsOnExit` means nothing. If some users want to keep the temporary files for n days as a backup even after the server restarts, they have to set both `MultipartRemovalStrategy.NEVER` and `purgeMultipartUploadsOnExit=true`.

>  that registers a shutdown hook that scans the directory and remove the items.

How about moving multipart files to a special directory such as `temporary` if `ON_JVM_SHUTDOWN` is set? The shutdown hook will delete all `temporary` directories under `multipartUploadsLocation`s.
```
- service1-upload/
  - incomplete/
  - running/ <- for ON_RESPONSE_COMPLETION
  - complete/ <- for NEVER
  - temporary/ <- for ON_JVM_SHUTDOWN
- service2-upload/
  - incomplete/
  - running/
  - temporary/
```

Reply:
I'm +1 for @ikhoon 's idea to at least group the user-facing APIs in one place. I do think we will need to add a shutdown hook somewhere if `ON_JVM_SHUTDOWN` is selected though.

```
val mps = MultipartRemovalStrategy.
  // builder methods
  .onJvmShutdown()
  .onResponseComplete()
  .never()
  // maybe advanced
  .builderForScheduledDeletion().poll(1 day).olderThan(7 days)
ServerBuilder.multipartRemovalStrategy(mps)
```

> How about moving multipart files to a special directory such as temporary if ON_JVM_SHUTDOWN is set?

I'm not sure I understood the need to separate the directory between `complete` and `temporary`.

Also, would it make sense to somehow embed the server id for each multipart file in case multiple servers are running in the same jvm? This way, we can still delete files that are for a specific jvm and the flag `ON_JVM_SHUTDOWN` makes more sense.

e.g. "<pid>-<server id>-" as the prefix for the temporary files.

Reply:
I'm unsure if `ON_JVM_SHUTDOWN` is a really useful option, it is worth implementing the complicated functions. For now, I would like to delete `ON_JVM_SHUTDOWN` from the PR and make it unsupported. If users have requests for it, we would like to consider them.

Reply:
> For now, I would like to delete `ON_JVM_SHUTDOWN` from the PR and make it unsupported.

Sounds good to me.